Build error caused by IS_J9_OBJECT_MONITOR_OWNER_DETACHED()

Building OpenJ9 Java 17 (and older) fails with the following error:

```
/home/openj9/openj9-openjdk-jdk17/openj9/runtime/util/moninfo.c: In function 'getObjectMonitorOwner':
/home/openj9/openj9-openjdk-jdk17/openj9/runtime/util/moninfo.c:89:38: error: implicit declaration of function 'IS_J9_OBJECT_MONITOR_OWNER_DETACHED'; did you mean 'J9_OBJECT_MONITOR_ENTER_FAILED'? [-Werror=implicit-function-declaration]
   89 |                                 if (!IS_J9_OBJECT_MONITOR_OWNER_DETACHED(osOwner)) {
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      J9_OBJECT_MONITOR_ENTER_FAILED
```

The macro was introduced by PR #21728, and the macro is defined only for Java 19 and above.

--- a/runtime/oti/j9nonbuilder.h
+++ b/runtime/oti/j9nonbuilder.h
@@ -5523,6 +5523,13 @@ typedef struct J9VMContinuation {
 } J9VMContinuation;
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if JAVA_SPEC_VERSION >= 24
+#define J9_OBJECT_MONITOR_OWNER_DETACHED 0x1
+#define IS_J9_OBJECT_MONITOR_OWNER_DETACHED(owner) (J9_OBJECT_MONITOR_OWNER_DETACHED == (UDATA)(owner))
+#else /* JAVA_SPEC_VERSION >= 24 */
+#define IS_J9_OBJECT_MONITOR_OWNER_DETACHED(owner) FALSE
+#endif /* JAVA_SPEC_VERSION >= 24 */
+
 /* @ddr_namespace: map_to_type=J9VMThread */
 
 typedef struct J9VMThread {

--- a/runtime/rasdump/javadump.cpp
+++ b/runtime/rasdump/javadump.cpp
@@ -4313,7 +4313,9 @@ JavaCoreDumpWriter::writeMonitorObject(J9ThreadMonitor* monitor, j9object_t obj,
 	if (NULL != obj) {
 		owner = getObjectMonitorOwner(_VirtualMachine, obj, &count);
 	} else if (NULL != lockOwner) {
-		owner = getVMThreadFromOMRThread(_VirtualMachine, lockOwner);
+		if (!IS_J9_OBJECT_MONITOR_OWNER_DETACHED(lockOwner)) {
+			owner = getVMThreadFromOMRThread(_VirtualMachine, lockOwner);
+		}
 		count = lock->count;
 	}
 
@@ -4338,7 +4340,16 @@ JavaCoreDumpWriter::writeMonitorObject(J9ThreadMonitor* monitor, j9object_t obj,
 	/* Describe its owning thread */
 	bool inflated = J9_ARE_ANY_BITS_SET(lock->flags, J9THREAD_MONITOR_INFLATED);
 
-	if ((NULL != owner) || (NULL != lockOwner)) {
+	if (IS_J9_OBJECT_MONITOR_OWNER_DETACHED(lockOwner)) {
+		if (inflated) {
+			_OutputStream.writeCharacters("owner \"");
+		} else {
+			_OutputStream.writeCharacters("Flat locked by \"");
+		}
+		_OutputStream.writeCharacters("<detached virtual thread>");
+		_OutputStream.writeCharacters(", entry count ");
+		_OutputStream.writeInteger(count, "%zu");
+	} else if ((NULL != owner) || (NULL != lockOwner)) {
 		if (inflated) {
 			_OutputStream.writeCharacters("owner \"");
 		} else {

--- a/runtime/util/moninfo.c
+++ b/runtime/util/moninfo.c
@@ -86,7 +86,9 @@ getObjectMonitorOwner(J9JavaVM *vm, j9object_t object, UDATA *pcount)
 		if (monitor) {
 			omrthread_t osOwner = monitor->owner;
 			if (osOwner) {
-				owner = getVMThreadFromOMRThread(vm, osOwner);
+				if (!IS_J9_OBJECT_MONITOR_OWNER_DETACHED(osOwner)) {
+					owner = getVMThreadFromOMRThread(vm, osOwner);
+				}
 				/* possible timing hole -- owner might exit monitor */
 				count = monitor->count;
 				if (count == 0) {

--- a/runtime/util/thrinfo.c
+++ b/runtime/util/thrinfo.c
@@ -264,7 +264,9 @@ getVMThreadStateHelper(J9VMThread *targetThread,
 							 * owned by a competing thread.
 							 */
 							vmstate = J9VMTHREAD_STATE_BLOCKED;
-							lockOwner = getVMThreadFromOMRThread(targetThread->javaVM, j9owner);
+							if (!IS_J9_OBJECT_MONITOR_OWNER_DETACHED(j9owner)) {
+								lockOwner = getVMThreadFromOMRThread(targetThread->javaVM, j9owner);
+							}
 							rawLock = (omrthread_monitor_t)objmon;
 						}
 					} else {
@@ -274,7 +276,9 @@ getVMThreadStateHelper(J9VMThread *targetThread,
 							} else {
 								vmstate = J9VMTHREAD_STATE_WAITING;
 							}
-							lockOwner = getVMThreadFromOMRThread(targetThread->javaVM, j9owner);
+							if (!IS_J9_OBJECT_MONITOR_OWNER_DETACHED(j9owner)) {
+								lockOwner = getVMThreadFromOMRThread(targetThread->javaVM, j9owner);
+							}
 							rawLock = (omrthread_monitor_t)objmon;
 							
 						} else if ((omrthread_monitor_t)objmon == j9state.blocker) {

--- a/runtime/vm/ContinuationHelpers.cpp
+++ b/runtime/vm/ContinuationHelpers.cpp
@@ -797,7 +797,7 @@ detachMonitorInfo(J9VMThread *currentThread, j9object_t lockObject)
 	}
 
 	J9ThreadAbstractMonitor *monitor = (J9ThreadAbstractMonitor *)objectMonitor->monitor;
-	monitor->owner = (J9Thread*)1;
+	monitor->owner = (J9Thread *)J9_OBJECT_MONITOR_OWNER_DETACHED;
 	objectMonitor->ownerContinuation = currentThread->currentContinuation;
 
 	return objectMonitor;
@@ -981,7 +981,7 @@ restart:
 					&& J9_ARE_NO_BITS_SET(continuation->runtimeFlags, J9VM_CONTINUATION_RUNTIMEFLAG_JVMTI_CONTENDED_MONITOR_ENTER_RECORDED)
 				) {
 					/* Get owner here since it could be too late within yieldPinnedContinuation(). */
-					if ((void*)1 == monitor->owner) {
+					if (IS_J9_OBJECT_MONITOR_OWNER_DETACHED(monitor->owner)) {
 						continuation->previousOwner = NULL;
 					} else {
 						continuation->previousOwner = getVMThreadFromOMRThread(vm, ((J9ThreadMonitor *)monitor)->owner);

--- a/runtime/vm/ObjectMonitor.cpp
+++ b/runtime/vm/ObjectMonitor.cpp
@@ -179,7 +179,7 @@ objectMonitorEnterBlocking(J9VMThread *currentThread)
 		 */
 		IDATA waitTime = 1;
 
-		if ((NULL != previousOMRThreadOwner) && (1 != (UDATA)previousOMRThreadOwner)) {
+		if ((NULL != previousOMRThreadOwner) && !IS_J9_OBJECT_MONITOR_OWNER_DETACHED(previousOMRThreadOwner)) {
 			previousOwner = getVMThreadFromOMRThread(vm, previousOMRThreadOwner);
 		}
 

--- a/runtime/vm/monhelpers.c
+++ b/runtime/vm/monhelpers.c
@@ -163,7 +163,7 @@ restart:
 		objectMonitor = J9_INFLLOCK_OBJECT_MONITOR(lock);
 		monitor = (J9ThreadAbstractMonitor *)objectMonitor->monitor;
 		Assert_VM_notNull(monitor);
-		Assert_VM_false(monitor->owner == (void*)1);
+		Assert_VM_false(IS_J9_OBJECT_MONITOR_OWNER_DETACHED(monitor->owner));
 
 #ifdef OMR_THR_ADAPTIVE_SPIN
 		/* for now we don't allow deflation if spinning has been disabled for this monitor

